time-series-utils/preprocessing/script.whizzml

```diff
@@ -0,0 +1,145 @@
+;; Given a dataset, extract its rows in a list format
+;; Input: ds-id - Dataset ID
+;; Output: list of rows values (no headers provided)
+(define (fetch-dataset-rows ds-id)
+  (let (sample-id (create-and-wait-sample {"dataset" ds-id})
+        rows ((fetch ds-id) "rows")
+        sample (fetch sample-id {"rows" rows
+                                 "mode" "linear"})
+        rows (sample ["sample" "rows"] [])
+        rows (for (row rows)
+               (for (item row)
+                    (if (not item)
+                        "NaN"
+                        item))))
+    (delete (wait sample-id))
+    rows))
+    
+    
+;; Checks for gaps in a epoch field and adds missing rows to ensure all
+;; dates are present. It expects the rows to be sorted previously in an
+;; ascendent sequence.
+;; Input:
+;;   rows - list of rows values
+;;   key-col - column that contains the ordering sequence datetime epoch
+;;   lapse - milliseconds to next row
+(define (fill-gaps rows key-col num-key-col lapse)
+  ((iterate (acc [-1 []] row rows)
+     (if (< (acc 0) 0)
+         [(row num-key-col)  (append (acc 1) row)]
+         (let (next (+ (acc 0) lapse))
+           (if (= next (row num-key-col))
+               [(row num-key-col) (append (acc 1) row)]
+               [(row num-key-col)
+                (append (concat (acc 1) (generate-missing-rows (count row) key-col num-key-col next (row num-key-col) lapse))
+                        row)])))) 1))
+                        
+;; Creates a list of missing rows of the given length with an epoch datetime
+;; key column whose value is generated in the range defined by the start, end
+;; and lapse parameters
+(define (generate-missing-rows row-length key-col num-key-col start end lapse)
+  (for (time (range start end lapse))
+       (for (col (range 0 row-length))
+            (if (= num-key-col col)
+                time
+                ""))))
+                
+                
+;; Keeps only the first row based on a key-col epoch field, thus ensuring that no
+;; duplicates will be found for the key field 
+(define (first rows num-key-col)
+  ((iterate (acc [-1 []] row rows)
+     (if (< (acc 0) 0)
+         [(row num-key-col) (append (acc 1) row)]
+         (if (= (row num-key-col) (acc 0))
+             acc
+             [(row num-key-col) (append (acc 1) row)]))) 1))
```

Comment:
here and above, i'd put the argument to the return value of iterate (the 1) in a line by its own, aligned with the second parenthesis of ((iterate, to make it easier to read.